openslides_backend/action/actions/user/user_scope_permission_check_mixin.py

```diff
@@ -0,0 +1,94 @@
+from enum import Enum, auto
+from typing import Any, Dict, Tuple
+
+from ....permissions.management_levels import (
+    CommitteeManagementLevel,
+    OrganisationManagementLevel,
+)
+from ....permissions.permission_helper import (
+    has_committee_management_level,
+    has_organisation_management_level,
+    has_perm,
+)
+from ....permissions.permissions import Permissions
+from ....services.datastore.interface import GetManyRequest
+from ....shared.exceptions import MissingPermission
+from ....shared.patterns import Collection, FullQualifiedId
+from ...action import Action
+
+
+class UserScope(Enum):
+    Meeting = auto()
+    Committee = auto()
+    Organisation = auto()
+
+
+class UserScopePermissionCheckMixin(Action):
+    def check_permissions_for_scope(self, instance: Dict[str, Any]) -> None:
+        """
+        Checks the permissions for user-altering actions depending on the user scope.
+        """
+        if has_organisation_management_level(
+            self.datastore, self.user_id, OrganisationManagementLevel.CAN_MANAGE_USERS
+        ):
+            return
+
+        scope, scope_id = self.get_user_scope(instance["id"])
+        if scope == UserScope.Committee:
+            if not has_committee_management_level(
+                self.datastore,
+                self.user_id,
+                CommitteeManagementLevel.CAN_MANAGE,
+                scope_id,
+            ):
+                raise MissingPermission(
+                    {
+                        OrganisationManagementLevel.CAN_MANAGE_USERS: 1,
+                        CommitteeManagementLevel.CAN_MANAGE: scope_id,
+                    }
+                )
+        elif scope == UserScope.Meeting:
+            meeting = self.datastore.fetch_model(
+                FullQualifiedId(Collection("meeting"), scope_id), ["committee_id"]
+            )
+            if not has_committee_management_level(
+                self.datastore,
+                self.user_id,
+                CommitteeManagementLevel.CAN_MANAGE,
+                meeting["committee_id"],
+            ) and not has_perm(
+                self.datastore, self.user_id, Permissions.User.CAN_MANAGE, scope_id
+            ):
+                raise MissingPermission(
+                    {
+                        OrganisationManagementLevel.CAN_MANAGE_USERS: 1,
+                        CommitteeManagementLevel.CAN_MANAGE: meeting["committee_id"],
+                        Permissions.User.CAN_CHANGE_OWN_PASSWORD: scope_id,
+                    }
+                )
+        else:
+            raise MissingPermission({OrganisationManagementLevel.CAN_MANAGE_USERS: 1})
+
+    def get_user_scope(self, id: int) -> Tuple[UserScope, int]:
```

Comment:
I don't understand, what this is for. What's about a user belonging to 2 committees? Committe permissions is not valid, without Organisation Level he has no permission at all?!

Reply:
No, as soon as the user is in 2 meetings, he is either in committee scope (if those belong to the same committee and the user is also part of that) or else in organisation scope. The use case is that if you can create a user for your scope, you should be able to edit it as well. As soon as the user is moved up in his scope by an admin with more permissions, you do not have permissions to edit it anymore. A user has never multiple scopes or, in this case, mutiple scope ids.

Reply:
> No, as soon as the user is in 2 meetings, he is either in committee scope (if those belong to the same committee and the user is also part of that) or else in organisation scope.
Imagine the request user has permission user.can_manage in meeting 1 and 2. The field group_$_ids belongs to field group C. With this permissions the request user is able to create a new user or update an existing user with groups for both meetings. See group permission [here](https://github.com/OpenSlides/OpenSlides/wiki/user.create#permissions).

I'm not sure about the idea of scope: Do you want to prevent the request user, even with permission for both meetings, to create such a user? Same for committees, one level above?
This would facilitate the permission handling for users a lot!

> The use case is that if you can create a user for your scope, you should be able to edit it as well. As soon as the user is moved up in his scope by an admin with more permissions, you do not have permissions to edit it anymore. A user has never multiple scopes or, in this case, multiple scope ids.

Reply:
For creating a user: If you have user.can_manage in two meetings of the same committe, but you cannot manage the committee, you must create two user instances for the same person. Since a user with two meetings would be in committee scope in this case. This is wanted - if you are a manager of multiple meetings in a committee, you should actually be a committee manager.

The more interesting thing is updating a user, especially groups B and C:

Currently, there should be [the basic check](https://github.com/OpenSlides/OpenSlides/wiki/Users#Permissions-for-altering-a-user) (I think it is the `check_permissions_for_scope` function, right?). For altering fields of group B it does not make any sense. If you have `user.can_manage` in a meeting and the updated user is part of the meeting, you should be able to change meeting-specific fields **regardless** of the scope. This is logical since mostly only the meeting adminscare about meeting-specific attributes. (Users with CML/OML can also update these fields, but must follow the scope guidelines)

Group C is even more tricky: The same argument as for group B holds: If there you have `user.can_manage` in a meeting, you can change the groups from users in your meeting. This has a speciality: You cannot set the groups for a user, that didn't had groups before (if you do not have CML/OML for managing users). In other words: You cannot put new users into the meeting. The same goes for setting the groups to `[]`, which must be set to `[<default_group_id>]` if the user only is a meeting admin.

All other fields, especially group A are protected by the basic check.

Reply:
| The same goes for setting the groups to [], which must be set to [<default_group_id>] if the user only is a meeting admin.

So the meeting admin should not be able to remove users from the meeting? This is still missing in https://github.com/OpenSlides/OpenSlides/wiki/user.update